### PR TITLE
Update feature_dism with cache

### PR DESCRIPTION
**Pull request by matthew-churcher**
Add caching to feature_dism see #333 for more detail

### CODE REVIEW

File: providers/feature_dism.rb [hunk 1]
```diff
@@ -69,3 +67,14 @@ def dism
     locate_sysnative_cmd('dism.exe')
   end
 end
+
+@@feature_list = nil
+def get_feature_list
```
Comment (mwrock) [inline providers/feature_dism.rb]: Might be nice to have this create a hash of features to avoid the parsing in the consuming methods

**Comment by mwrock** [comment]
Good to see this in `feature_dism`!

This came up before but I can see cases where consuming multiple windows based cookbooks, you will have recipes possibly checking features via different providers. So one could run into this sequence:
1. Hydrate the dism cache
2. Install a feature via powershell
3. Try to install the same feature here later in the run list.

I **think** this is an edge case and would likely just result in an unnecessary call to install the feature but not result in an error. So I think this is ok but if anyone sees a graver danger we should address it.

**Comment by matthew-churcher** [comment]
@mwrock I made the suggested change, a hash is now used to store the features to avoid the parsing dism output again.

File: providers/feature_dism.rb [hunk 2]
```diff
@@ -69,3 +68,22 @@ def dism
     locate_sysnative_cmd('dism.exe')
   end
 end
+@@feature_list = nil
+def get_feature_list
+  if @@feature_list.nil?
+    list = {}
+    out = shell_out("#{dism} /online /Get-Features", returns: [0, 42, 127]).stdout
```
Comment (mwrock) [inline providers/feature_dism.rb]: should this be `feature_matches` instead of `out`?

File: providers/feature_dism.rb [hunk 3]
```diff
@@ -69,3 +68,22 @@ def dism
...
+    feature_matches.each do |match|
```
Comment (mwrock) [inline providers/feature_dism.rb]: could use:
`@@feature_list = feature_matches.each_with_object({}) do |match, list|`

File: providers/feature_dism.rb (hunk 2 above)
Comment (matthew-churcher) [inline providers/feature_dism.rb]: Nice, spot. There is an entire line missing. Not sure how that made it through testing.

File: providers/feature_dism.rb (hunk 3 above)
Comment (matthew-churcher) [inline providers/feature_dism.rb]: done

File: providers/feature_dism.rb [hunk 4]
```diff
@@ -69,3 +68,22 @@ def dism
...
+  if @@feature_list.nil?
+    out = shell_out("#{dism} /online /Get-Features", returns: [0, 42, 127]).stdout
+    feature_matches = out.scan(/^Feature Name : (.*).?$\n^State : (.*)?$/)
+    list = feature_matches.each_with_object({}) do |match, list|
+      name = match[0]
+      state = match[1]
+      state = 'Removed' if state.include? 'Payload Removed'
+      list[name] = state.to_sym
+    end
+    @@feature_list = list
```
Comment (mwrock) [inline providers/feature_dism.rb]: just a nit: I'd lose this line and just assign `@@feature_list` to the block

**Comment by mwrock** [comment]
:+1:

**Comment by matthew-churcher** [comment]
@mwrock Thanks for the feed back. It's been a while since ruby was my goto language! Technically I'm a performance test engineer :) probably shouldn't be writing code at all, lol.

What are the prospects of getting this merged? the team at NVM is keen to get a speed up in this area but we would prefer not to maintain a fork. This patch speed's up each of our converges by several minutes. (yay)

**Comment by matthew-churcher** [comment]
Oh deer, says I closed it. How to I reopen it?

**Comment by mwrock** [comment]
cc @chef-cookbooks/windows-cookbook-collaborators @chef-cookbooks/client-windows

**Comment by tas50** [comment]
The latest release of this cookbook includes caching via an Ohai plugin for dism. This should achieve the same thing and gives us a nice cache reloading method within ohai.

**Pull request closed without merging**